sklearn/_build_utils/cythonize.py

```diff
@@ -0,0 +1,192 @@
+import subprocess
+
+HASH_FILE = 'cythonize.dat'
+DEFAULT_ROOT = 'sklearn'
+
+# WindowsError is not defined on unix systems
+try:
+    WindowsError
+except NameError:
+    WindowsError = None
+
+
+#
+# Rules
+#
+def process_pyx(fromfile, tofile):
+    try:
+        from Cython.Compiler.Version import version as cython_version
+        from distutils.version import LooseVersion
+        if LooseVersion(cython_version) < LooseVersion('0.21'):
+            raise Exception('Building scikit-learn requires Cython >= 0.21')
+
+    except ImportError:
+        pass
+
+    flags = ['--fast-fail']
+    if tofile.endswith('.cpp'):
+        flags += ['--cplus']
+
+    try:
+        try:
+            r = subprocess.call(['cython'] + flags + ["-o", tofile, fromfile])
+            if r != 0:
+                raise Exception('Cython failed')
+        except OSError:
+            # There are ways of installing Cython that don't result in a cython
+            # executable on the path, see gh-2397.
+            r = subprocess.call([sys.executable, '-c',
+                                 'import sys; from Cython.Compiler.Main '
+                                 'import setuptools_main as main;'
+                                 ' sys.exit(main())'] + flags +
+                                ["-o", tofile, fromfile])
+            if r != 0:
+                raise Exception('Cython failed')
+    except OSError:
+        raise OSError('Cython needs to be installed')
+
+
+rules = {
+    '.pyx': process_pyx,
+}
+
+
+#
+# Hash db
+#
+def load_hashes(filename):
+    # Return { filename : (sha1 of input, sha1 of output) }
+    if os.path.isfile(filename):
+        hashes = {}
+        with open(filename, 'r') as f:
+            for line in f:
+                filename, inhash, outhash = line.split()
+                hashes[filename] = (inhash, outhash)
+    else:
+        hashes = {}
+    return hashes
+
+
+def save_hashes(hash_db, filename):
+    with open(filename, 'w') as f:
+        for key, value in sorted(hash_db.items()):
+            f.write("%s %s %s\n" % (key, value[0], value[1]))
+
+
+def sha1_of_file(filename):
+    h = hashlib.sha1()
+    with open(filename, "rb") as f:
+        h.update(f.read())
+    return h.hexdigest()
+
+
+#
+# Main program
+#
+
+def normpath(path):
+    path = path.replace(os.sep, '/')
+    if path.startswith('./'):
+        path = path[2:]
+    return path
+
+
+def get_hash(frompath, topath):
+    from_hash = sha1_of_file(frompath)
+    to_hash = sha1_of_file(topath) if os.path.exists(topath) else None
+    return (from_hash, to_hash)
+
+
+def process(path, fromfile, tofile, processor_function, hash_db):
+    fullfrompath = os.path.join(path, fromfile)
+    fulltopath = os.path.join(path, tofile)
+    current_hash = get_hash(fullfrompath, fulltopath)
+    if current_hash == hash_db.get(normpath(fullfrompath), None):
+        print('%s has not changed' % fullfrompath)
+        return
+
+    orig_cwd = os.getcwd()
+    try:
```

Comment:
I assume this is possible to do, without `chdir`ing to and fro by providing the `fullfrompath` and `fulltopath` directly.